garage/tf/algos/reps.py

```diff
@@ -0,0 +1,442 @@
+
+            dual_loss += self.l2_reg_dual * (
+                tf.square(i.param_eta) + tf.square(1 / i.param_eta))
+
+            dual_grad = tf.gradients(dual_loss, [i.param_eta, i.param_v])
+
+        self.f_dual = tensor_utils.compile_function(
+            flatten_inputs(self._dual_opt_inputs),
+            dual_loss,
+            log_name="f_dual")
+
+        self.f_dual_grad = tensor_utils.compile_function(
+            flatten_inputs(self._dual_opt_inputs),
+            dual_grad,
+            log_name="f_dual_grad")
+
+        self.f_policy_kl = tensor_utils.compile_function(
+            flatten_inputs(self._policy_opt_inputs),
+            pol_mean_kl,
+            log_name="f_policy_kl")
+
+        return loss
+
+    def _dual_opt_input_values(self, samples_data):
+        """Update dual func optimize input values based on samples data."""
+        policy_state_info_list = [
+            samples_data["agent_infos"][k]
+            for k in self.policy.state_info_keys
+        ]   # yapf: disable
+        policy_old_dist_info_list = [
+            samples_data["agent_infos"][k]
+            for k in self.policy.distribution.dist_info_keys
+        ]
+
+        self.feat_diff = self._features(samples_data)
+
+        dual_opt_input_values = self._dual_opt_inputs._replace(
+            reward_var=samples_data["rewards"],
+            valid_var=samples_data["valids"],
+            feat_diff=self.feat_diff,
+            param_eta=self.param_eta,
+            param_v=self.param_v,
+            policy_state_info_vars_list=policy_state_info_list,
+            policy_old_dist_info_vars_list=policy_old_dist_info_list,
+        )
+
+        return flatten_inputs(dual_opt_input_values)
+
+    def _policy_opt_input_values(self, samples_data):
+        """Update policy optimize input values based on samples data."""
+        policy_state_info_list = [
+            samples_data["agent_infos"][k]
+            for k in self.policy.state_info_keys
+        ]   # yapf: disable
+        policy_old_dist_info_list = [
+            samples_data["agent_infos"][k]
+            for k in self.policy.distribution.dist_info_keys
+        ]
+
+        policy_opt_input_values = self._policy_opt_inputs._replace(
+            obs_var=samples_data["observations"],
+            action_var=samples_data["actions"],
+            reward_var=samples_data["rewards"],
+            valid_var=samples_data["valids"],
+            feat_diff=self.feat_diff,
+            param_eta=self.param_eta,
+            param_v=self.param_v,
+            policy_state_info_vars_list=policy_state_info_list,
+            policy_old_dist_info_vars_list=policy_old_dist_info_list,
+        )
+
+        return flatten_inputs(policy_opt_input_values)
+
+    def _features(self, samples_data):
+        """Get valid view features based on samples data."""
+        paths = samples_data["paths"]
+        feat_diff = []
+        for path in paths:
+            o = np.clip(path["observations"], -10, 10)
```

Comment:
Why is clipping min/max set to -10/10?

Reply:
this is worth a comment. i'm pretty sure it's to limit the dynamic range of the observations.

Reply:
I couldn't find how to compute features in the paper, so this method is ported from theano/algos/reps.py.

Reply:
I agree this is to limit the range of the observations. I don't know why they do it here instead of the sampler, so that we could normalize the observations, like Zhanpeng proposed. Another constant param is [100.0 in here](https://github.com/rlworkgroup/garage/pull/471/files#diff-1bc9d53a38cba8890b10755ab1366305R435). Is it related to max_path_length or max_reward?